File: simulator/src/message/Message.cpp

#include "Message.h"

Message::Message()
{
}

Message::Message( int dataLength, int type, int id, int sequenceId, int version )
{
    header = new MessageHeader( MessageHeader::getHeaderSize() + dataLength + MessageTrailer::getTrailerSize(),
                                type, id, sequenceId, version );
    trailer = new MessageTrailer();
}

Message::Message( int dataLength, int type, int id, int version )
{
    header = new MessageHeader( MessageHeader::getHeaderSize() + dataLength + MessageTrailer::getTrailerSize(),
                                type, id, version );
    trailer = new MessageTrailer();
}

Message::~Message()
{
    if( trailer != nullptr )
    {
        delete trailer;
    }

    if( header != nullptr )
    {
        delete header;
    }
}

MessageHeader *
Message::getHeader()
{
    return header;
}

MessageTrailer *
Message::getTrailer()
{
    return trailer;
}

void
Message::moveFunction( MessageProcessor *messageProcessor,
                       int direction, unsigned int amount )
{
    messageProcessor->moveFunction( direction, amount );
}

void
Message::stopFunction( MessageProcessor *messageProcessor )
{
    messageProcessor->stopFunction();
}

File: interface/src/com/christopher/robot/message/Message.java
package com.christopher.robot.message;

import java.nio.ByteBuffer;

public abstract class Message
{
    protected MessageHeader header;
    protected MessageTrailer trailer;

    public abstract Response processMessage();
    public abstract void pack( ByteBuffer buffer );

    protected Message()
    {
    }

    protected Message( int dataLength, int type, int id, int sequenceId,
                       int version )
    {
        header = new MessageHeader( MessageHeader.getSize() + dataLength + MessageTrailer.getSize(),
                                    type, id, sequenceId, version );
        trailer = new MessageTrailer();
    }

    protected Message( int dataLength, int type, int id, int version )
    {
        header = new MessageHeader( MessageHeader.getSize() + dataLength + MessageTrailer.getSize(),
                                    type, id, version );
        trailer = new MessageTrailer();
    }

    public MessageHeader getHeader()
    {
        return header;
    }

    public MessageTrailer trailer()
    {
        return trailer;
    }

    public int getLength()
    {
        int length = 0;

        if( header != null )
        {
            length = header.getLength();
        }

        return length;
    }
}
